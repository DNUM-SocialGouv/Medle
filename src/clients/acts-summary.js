import { ACTS_SUMMARY_ENDPOINT, API_URL } from "../config"
import { handleAPIResponse2 } from "../utils/errors"

export const findSummaryByHospital = async ({  hospitalId, headers }) => {
    const response = await fetch(`${API_URL}${ACTS_SUMMARY_ENDPOINT}?hospitals=${hospitalId}`, {
      headers: { ...headers, "Content-Type": "application/json" },
    })
  
    const { elements } = await handleAPIResponse2(response)
    return elements
  }

  export const fetchExport = async (selectedYear, hospitalId, headers) => {
    saveAs(`${API_URL}${ACTS_SUMMARY_ENDPOINT}/export?hospitals=${hospitalId}&year=${selectedYear}`, { ...headers })
  }
