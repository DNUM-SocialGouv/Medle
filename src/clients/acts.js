import { saveAs } from "file-saver"
import fetch from "isomorphic-unfetch"

import { ACTS_ENDPOINT, API_URL } from "../config"
import { handleAPIResponse } from "../utils/errors"
import { METHOD_DELETE, METHOD_POST, METHOD_PUT } from "../utils/http"

export const findAct = async ({ id, headers }) => {
  const response = await fetch(API_URL + ACTS_ENDPOINT + "/" + id, { headers })
  return handleAPIResponse(response)
}

export const searchActsByKey = async ({ key, value, headers }) => {
  const response = await fetch(`${API_URL}${ACTS_ENDPOINT}?${key}=${value}`, {
    headers: { ...headers, "Content-Type": "application/json" },
  })

  const { elements } = await handleAPIResponse(response)
  return elements
}

const buildQueryParams = (params) => {
  const { search, startDate, endDate, hospitals, profiles, asker, requestedPage } = params

  const arr = []
  if (search) {
    arr.push(`fuzzy=${search}`)
  }
  if (startDate) {
    arr.push(`startDate=${startDate}`)
  }
  if (endDate) {
    arr.push(`endDate=${endDate}`)
  }
  if (hospitals?.length) {
    arr.push(`hospitals=${hospitals.map((elt) => elt.value).join(",")}`)
  }
  if (profiles?.length) {
    arr.push(`profiles=${profiles.map((elt) => elt.value).join(",")}`)
  }
  if (asker?.value) {
    arr.push(`asker=${asker?.value}`)
  }
  if (requestedPage) {
    arr.push(`requestedPage=${requestedPage}`)
  }
  return arr.length ? "?" + arr.join("&") : ""
}

export const searchActs = async (params) => {
  const { headers } = params

  const queryParams = buildQueryParams(params)

  const response = await fetch(`${API_URL}${ACTS_ENDPOINT}${queryParams}`, { headers })

  return handleAPIResponse(response)
}

export const createAct = async ({ act, headers }) => {
  const response = await fetch(API_URL + ACTS_ENDPOINT, {
    method: METHOD_POST,
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify(act),
  })
  return handleAPIResponse(response)
}

export const updateAct = async ({ act, headers }) => {
  const response = await fetch(API_URL + ACTS_ENDPOINT + "/" + act.id, {
    method: METHOD_PUT,
    headers: { ...headers, "Content-Type": "application/json" },
    body: JSON.stringify(act),
  })
  return handleAPIResponse(response)
}

export const deleteAct = async ({ id, headers }) => {
  const response = await fetch(`${API_URL}${ACTS_ENDPOINT}/${id}`, { method: METHOD_DELETE, headers })
  return handleAPIResponse(response)
}

export const fetchExport = async (params) => {
  const { headers } = params

  const queryParams = buildQueryParams(params)

  saveAs(`${API_URL}${ACTS_ENDPOINT}/export${queryParams}`, { headers })
}
