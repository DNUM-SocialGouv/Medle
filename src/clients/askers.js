import fetch from "isomorphic-unfetch"
import moize from "moize"

import { API_URL, ASKERS_ENDPOINT } from "../config"
import { handleAPIResponse } from "../utils/errors"
import { METHOD_POST, METHOD_PUT, METHOD_DELETE } from "../utils/http"
import { isEmpty } from "../utils/misc"

const MAX_AGE = 1000 * 60 * 60 // 1 hour

export const searchAskersFuzzy = async ({ search, all = false, headers }) => {
   search = search ? [`fuzzy=${search}`] : []
   all = all ? ["all=true"] : []

   let bonus = [...search, ...all]
   bonus = bonus.length ? `?${bonus.join("&")}` : ""

   const response = await fetch(`${API_URL}${ASKERS_ENDPOINT}${bonus}`, { headers })
   const json = await handleAPIResponse(response)
   return isEmpty(json) ? [] : json
}

export const findAsker = async ({ id, headers }) => {
   const response = await fetch(`${API_URL}${ASKERS_ENDPOINT}/${id}`, { headers })
   return handleAPIResponse(response)
}

export const memoizedSearchAskers = moize({ maxAge: MAX_AGE, isPromise: false })(searchAskersFuzzy)
export const memoizedFindAsker = moize({ maxAge: MAX_AGE, isDeepEqual: true })(findAsker)

export const createAsker = async ({ asker, headers }) => {
   const response = await fetch(`${API_URL}${ASKERS_ENDPOINT}`, {
      method: METHOD_POST,
      headers: { ...headers, "Content-Type": "application/json" },
      body: JSON.stringify(asker),
   })

   return handleAPIResponse(response)
}

export const updateAsker = async ({ asker, headers }) => {
   const response = await fetch(`${API_URL}${ASKERS_ENDPOINT}/${asker.id}`, {
      method: METHOD_PUT,
      headers: { ...headers, "Content-Type": "application/json" },
      body: JSON.stringify(asker),
   })
   return handleAPIResponse(response)
}

export const deleteAsker = async ({ id, headers }) => {
   const response = await fetch(`${API_URL}${ASKERS_ENDPOINT}/${id}`, { method: METHOD_DELETE, headers })
   return handleAPIResponse(response)
}
