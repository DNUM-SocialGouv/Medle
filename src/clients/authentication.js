import fetch from "isomorphic-unfetch"

import { API_URL, LOGIN_ENDPOINT } from "../config"
import { handleAPIResponse2 } from "../utils/errors"
import { METHOD_POST } from "../utils/http"

export const authenticate = async (email, password) => {
  const response = await fetch(API_URL + LOGIN_ENDPOINT, {
    method: METHOD_POST,
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email, password }),
  })

  const tokenCookie = response.headers.get("Set-Cookie") // may be null in CSR mode or because of CORS or secure cookie. OK in SSR and good for us for testing purpose.
  
  const { user, token } = await handleAPIResponse2(response)
  
  if (!user) {
    throw new Error("Authentication failed")
  }
  
  return {
    headers: {
      cookie: tokenCookie,
    },
    user,
    token
  }
}
