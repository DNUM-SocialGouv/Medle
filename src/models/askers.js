import * as yup from "yup"
import * as common from "./common"

const JStoDBKeys = {
  id: "id",
  name: "name",
  depCode: "dep_code",
  type: "type",
}

const schema = yup.object().shape({
  id: yup.number().positive().integer().nullable(),
  name: yup.string(),
  depCode: yup.string().matches(/^$|[0-9]{2,3}/),
  type: yup.string(),
})

export const { transform, transformAll, untransform, validate } = common.build({ JStoDBKeys, schema })
