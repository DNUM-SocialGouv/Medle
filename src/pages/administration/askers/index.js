import AddIcon from "@material-ui/icons/Add"
import Link from "next/link"
import { PropTypes } from "prop-types"
import React, { useState } from "react"
import { Alert, Col, Container, Form, FormGroup, Input, Spinner, Table } from "reactstrap"

import { searchAskersFuzzy } from "../../../clients/askers"
import { SearchButton } from "../../../components/form/SearchButton"
import Layout from "../../../components/Layout"
import Pagination from "../../../components/Pagination"
import { Title1 } from "../../../components/StyledComponents"
import { usePaginatedData } from "../../../hooks/usePaginatedData"
import { buildAuthHeaders, redirectIfUnauthorized, withAuthentication } from "../../../utils/auth"
import { logError } from "../../../utils/logger"
import { ADMIN } from "../../../utils/roles"

const AdminAskerPage = ({ paginatedData: initialPaginatedData, currentUser }) => {
  const [search, setSearch] = useState("")
  const [paginatedData, error, loading, fetchPage] = usePaginatedData(searchAskersFuzzy, initialPaginatedData)

  const onChange = (e) => {
    setSearch(e.target.value)
  }

  const onSubmit = async (e) => {
    e.preventDefault()
    fetchPage({ search })(0)
  }

  return (
    <Layout page="askers" currentUser={currentUser} admin={true}>
      <Container
        style={{ maxWidth: 980, minWidth: 740 }}
        className="mt-5 mb-5 d-flex justify-content-between align-items-baseline"
      >
        <Title1 className="">{"Administration des demandeurs"}</Title1>
        <Link href="/administration/askers/[id]" as={`/administration/askers/new`}>
          <a>
            <SearchButton className="btn-outline-primary">
              <AddIcon />
              &nbsp; Ajouter
            </SearchButton>
          </a>
        </Link>
      </Container>

      <Container style={{ maxWidth: 980, minWidth: 740 }}>
        <Form onSubmit={onSubmit}>
          <FormGroup row inline className="mb-4 justify-content-center">
            <Col className="flex-grow-1">
              <Input
                type="text"
                name="es"
                id="es"
                placeholder="Rechercher un demandeur par son nom, etc."
                value={search}
                onChange={onChange}
                autoComplete="off"
              />
            </Col>
            <Col className="flex-grow-0">
              <SearchButton className="btn-primary" disabled={loading} onClick={onSubmit}>
                {loading ? <Spinner size="sm" color="light" data-testid="loading" /> : "Chercher"}
              </SearchButton>
            </Col>
          </FormGroup>
        </Form>
        {error && (
          <Alert color="danger" className="mb-4">
            {error}
          </Alert>
        )}

        {!error && !paginatedData.elements.length && <div className="text-center">{"Aucun demandeur trouvé."}</div>}

        {!error && !!paginatedData.elements.length && (
          <>
            <Pagination data={paginatedData} fn={fetchPage(search)} />

            <Table responsive className="table-hover">
              <thead>
                <tr className="table-light">
                  <th>Nom</th>
                  <th>Département</th>
                </tr>
              </thead>
              <tbody>
                {paginatedData.elements.map((asker) => (
                  <Link key={asker.id} href="/administration/askers/[id]" as={`/administration/askers/${asker.id}`}>
                    <tr>
                      <td>
                        <b>{`${asker.name}`}</b>
                      </td>
                      <td>{asker.depCode}</td>
                      <td>
                        <Link href="/administration/askers/[id]" as={`/administration/askers/${asker.id}`}>
                          <a className="text-decoration-none">Voir</a>
                        </Link>
                      </td>
                    </tr>
                  </Link>
                ))}
              </tbody>
            </Table>
          </>
        )}
      </Container>
    </Layout>
  )
}

AdminAskerPage.getInitialProps = async (ctx) => {
  const headers = buildAuthHeaders(ctx)

  try {
    const paginatedData = await searchAskersFuzzy({ headers })
    return { paginatedData }
  } catch (error) {
    logError("APP error", error)

    redirectIfUnauthorized(error, ctx)
  }
  return {}
}

AdminAskerPage.propTypes = {
  paginatedData: PropTypes.object,
  currentUser: PropTypes.object.isRequired,
}

export default withAuthentication(AdminAskerPage, ADMIN)
