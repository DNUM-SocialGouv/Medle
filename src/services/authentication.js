import knex from "../knex/knex"
import { transform } from "../models/users"
import { compareWithHash } from "../utils/bcrypt"
import { APIError } from "../utils/errors"
import { STATUS_400_BAD_REQUEST, STATUS_401_UNAUTHORIZED } from "../utils/http"
import { generateToken } from "../utils/jwt"

const validPassword = (password) => !!password?.length

export const authenticate = async (email, password) => {
  // normalize email
  email = email?.trim()
  // request verification
  if (!validPassword(password)) {
    throw new APIError({
      status: STATUS_400_BAD_REQUEST,
      message: "Incorrect password",
    })
  }

  // SQL query
  const [dbUser] = await knex("users")
    .leftJoin("hospitals", "users.hospital_id", "hospitals.id")
    .where(knex.raw(`lower(email) = ?`, email?.toLowerCase()))
    .whereNull("users.deleted_at")
    .select(
      "users.id",
      "users.first_name",
      "users.last_name",
      "users.email",
      "users.password",
      "users.role",
      "users.hospital_id",
      "hospitals.name as hospital_name",
      "hospitals.extra_data as hospital_extra_data",
      "users.scope"
    )

  if (dbUser && (await compareWithHash(password, dbUser.password))) {
    const user = transform(dbUser)
    const token = generateToken(user)
    return { user, token }
  } else {
    // Unauthorized path
    throw new APIError({
      status: STATUS_401_UNAUTHORIZED,
      message: "Erreur d'authentification",
    })
  }
}
