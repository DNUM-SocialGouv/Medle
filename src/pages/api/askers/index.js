import Cors from "micro-cors"

import { STATUS_200_OK, METHOD_GET, METHOD_OPTIONS, METHOD_POST } from "../../../utils/http"
import { ADMIN, ACT_CONSULTATION } from "../../../utils/roles"
import { sendAPIError, sendMethodNotAllowedError } from "../../../services/errorHelpers"
import { checkIsSuperAdmin, checkValidUserWithPrivilege } from "../../../utils/auth"
import { create, search } from "../../../services/askers"

const handler = async (req, res) => {
  res.setHeader("Content-Type", "application/json")

  try {
    switch (req.method) {
      case METHOD_GET: {
        const currentUser = checkValidUserWithPrivilege(ACT_CONSULTATION, req, res)

        const { askers, totalCount, currentPage, maxPage, byPage } = await search({ ...req.query, currentUser })

        return res.status(STATUS_200_OK).json({ totalCount, currentPage, maxPage, byPage, elements: askers })
      }
      case METHOD_POST: {
        const currentUser = checkValidUserWithPrivilege(ADMIN, req, res)

        checkIsSuperAdmin(currentUser)

        const id = await create(req.body)

        return res.status(STATUS_200_OK).json({ id })
      }

      default:
        if (req.method !== METHOD_OPTIONS) return sendMethodNotAllowedError(res)
    }
  } catch (error) {
    sendAPIError(error, res)
  }
}

const cors = Cors({
  allowMethods: [METHOD_GET, METHOD_OPTIONS],
})

export default cors(handler)
