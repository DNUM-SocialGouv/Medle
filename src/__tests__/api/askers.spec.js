import { authenticate } from "../../clients/authentication"
import { searchAskersFuzzy } from "../../clients/askers"

const headersActUserTours = () => authenticate("acte@tours.fr", "test")

describe("/askers", () => {
  it("should return all criminal courts in France", async () => {
    const { headers } = await headersActUserTours()

    const askers = await searchAskersFuzzy({ search: "Montargis", headers })

    expect(askers).toMatchSnapshot()
  })
})
