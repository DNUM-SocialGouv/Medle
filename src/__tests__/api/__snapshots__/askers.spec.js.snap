// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/askers should return all criminal courts in France 1`] = `
Array [
  Object {
    "depCode": null,
    "id": 3028,
    "name": "Brigade de gendarmerie - Saint-Aignan",
    "type": "gendarmerie",
  },
  Object {
    "depCode": null,
    "id": 2846,
    "name": "Brigade de gendarmerie - Saint-Aignan-sur-RoÃ«",
    "type": "gendarmerie",
  },
]
`;
